src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+
+    def is_indexing_by_release(self) -> bool:
+        # An identifier is indexing by release if release is set.
+        return bool(self.release)
+
+    def to_indexing_by_debug_id(self) -> FlatFileIdentifier:
+        return FlatFileIdentifier(project_id=self.project_id, release=NULL_STRING, dist=NULL_STRING)
+
+
+def _generate_flat_file_indexing_cache_key(identifier: FlatFileIdentifier):
+    # The {1} is a key which is used to keep all the keys on the same Redis instance, in order to provide consistency
+    # guarantees for the atomic check of the state.
+    return (
+        "{1}:flat_file_indexing:%s"
+        % hashlib.sha1(
+            b"%s|%s|%s"
+            % (
+                str(identifier.project_id).encode(),
+                str(identifier.release).encode(),
+                str(identifier.dist).encode(),
+            )
+        ).hexdigest()
+    )
+
+
+def set_flat_files_being_indexed_if_null(identifiers: List[FlatFileIdentifier]) -> bool:
+    redis_client = get_redis_cluster_for_artifact_bundles()
+    cache_keys = [_generate_flat_file_indexing_cache_key(identifier) for identifier in identifiers]
+
+    pipeline = redis_client.pipeline()
+
+    # We want to monitor all the keys, so that in case there are changes, the system will abort the transaction.
+    pipeline.watch(*cache_keys)
+
+    # Check if both keys are null.
+    all_keys_none = all(pipeline.get(cache_key) is None for cache_key in cache_keys)
+    if not all_keys_none:
+        return False
+
+    result = True
+    try:
+        # The transaction is only started if the keys that were previously marked as none, haven't been touched.
+        pipeline.multi()
+
+        for cache_key in cache_keys:
+            pipeline.set(cache_key, 1, ex=INDEXING_CACHE_TIMEOUT)
+
+        pipeline.execute()
+    except Exception:
+        result = False
+        metrics.incr("artifact_bundle_flat_file_indexing.indexing_conflict")
+
+    # Reset the watched keys.
+    pipeline.unwatch()
+
+    return result
+
+
+def index_bundle_in_flat_file(
+    artifact_bundle: ArtifactBundle, identifier: FlatFileIdentifier
+) -> FlatFileIndexingState:
+    with ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=True) as archive:
+        identifiers = [identifier]
+
+        # In case we have an identifier that is indexed by release and the bundle has debug ids, we want to perform
+        # indexing on also the debug ids bundle.
+        if identifier.is_indexing_by_release() and archive.has_debug_ids():
+            identifiers.append(identifier.to_indexing_by_debug_id())
+
+        # Before starting, we have to make sure that no other identifiers are being indexed.
+        if not set_flat_files_being_indexed_if_null(identifiers):
+            return FlatFileIndexingState.CONFLICT
+
+        # For each identifier we now have to compute the index.
+        for identifier in identifiers:
+            index = FlatFileIndex.build(FlatFileIndexStore(identifier))
```

Comment:
This might be the place where all the backing store related logic can be, and where all the error handling and metrics should happen as well.